api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -617,112 +597,58 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
   }
 
   @Override
-  public ResponseEntity<FileDetail> renameNotebook(String workspace, String workspaceName,
-      NotebookRename rename) {
-    String newName = rename.getNewName();
-    if (!newName.matches("^.+\\.ipynb")) {
-      newName = newName + ".ipynb";
-    }
-    FileDetail fileDetail = notebookCloneOperation(workspace, workspaceName, rename.getName(),
-        newName);
-    notebookDeleteOperation(workspace, workspaceName, rename.getName());
-    return ResponseEntity.ok(fileDetail);
-  }
-
-  @Override
-  public ResponseEntity<FileDetail> cloneNotebook(String workspace, String workspaceName,
-      String notebookName) {
-    String newName = notebookName.replaceAll("\\.ipynb", " ") + "Clone.ipynb";
-    FileDetail fileDetail = notebookCloneOperation(workspace, workspaceName, notebookName, newName);
-    return ResponseEntity.ok(fileDetail);
+  public ResponseEntity<List<FileDetail>> getNoteBookList(String workspaceNamespace,
+      String workspaceId) {
+    List<FileDetail> fileList = notebooksService.getNotebooks(workspaceNamespace, workspaceId);
```

Comment:
Is there a broader design reason to move all of the logic out of the controller? A lot of the code that was here is pretty tightly married to the details of the API specification. AFAICT  the newly added NotebooksService effectively duplicates the API controller interface, which is already autogenerated for us via Swagger.

I can maybe buy it if this method is being reused from multiple locations or if there is high complexity, but I'm not seeing much of an advantage in these cases. In general I would prioritize test coverage of the actual API method behavior over testing of a service.

Reply:
Yes. this is something I feel pretty strongly about. I can prepare a larger presentation on this topic but as a quick rundown... Even if we do not test at the Service level, I think there are a few reasons to break logic out of the controllers.

1. Our controllers are too big which make them hard to navigate and also understand the responsibilities for each.
Ex. whats the difference in responsibilities between WorkspaceController and WorkspaceService? ConceptSetController and ConceptSetService?

2. Encourage layered architecture
For example, we can set the controllers as the layer between our business logic and HTTP and have it handle everything related to HTTP such as request/response parsing, as well as extracting state out of the request so the services can be stateless. (Grabbing User out of Provider<User>)

3. Enable code reuse and exportability
Related to #2: if we create a business logic layer that is purely Java and not tied to Spring, it enables us to extract libraries and jars that can be used in other repositories such as Terra. 
Of course, it also enables our own controllers and services to reuse code with less work.

4. Enable unit testing
I think there is another discussion to be had on whether or not we fully embrace unit testing but I think we should at least we designing our code in a way that it is possible.

My reasoning for pulling NotebookService out of WorkspaceController was that I identified a chunk of logic that was in WorkspaceController but had very little interaction with the rest of the controller. In my mind, WorkspaceController was effectively playing two roles of servicing Workspace related API calls and Notebooks related API alls. I think we should actually take things a step further and create a NotebooksController which handles notebooks related API calls. This work is a step in that direction.

Reply:
Thanks - glad to see some strong opinions here. I think it would a useful topic for next week's tech talk if you want to present this and (if possible) outline the before/after of the dependency graph / codebase structure. Seems fine to proceed with the approach in your current PR in the meantime.

Two points that I will add here to consider:
- We control which methods go in which controller. This is defined by the tag in our yaml file. We could decide to instead break up large controllers, e.g. place all the methods you've refactored here into their own controller (the interface would look the same, or perhaps identical to, the NotebooksService). This would also change the client interfaces, so we'd need to think about whether that would actually be desirable overall. My guess is the current distribution of methods could be improved significantly.
- Provider<User> is a request-context-driven injection of data. If your service is making remote calls, e.g. to Leo, it is also going to be reading this context information out, i.e. to authenticate the backend call. If the layer above and below a service are both pulling the User out of the context in this way, having a stateless service between them is less compelling - though I can buy there are still some benefits.

Reply:
Good points! I'd be happy to prepare a tech talk on this.